torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -249,8 +266,12 @@ void Reducer::autograd_hook(VariableIndex index) {
   // output, they won't be part of the autograd graph, and won't receive
   // gradients. These parameters are discovered in the `prepare_for_backward`
   // function and their indexes stored in the `unused_parameters_` vector.
-  if (!has_marked_unused_parameters_ && !unused_parameters_.empty()) {
+  if (!has_marked_unused_parameters_) {
     has_marked_unused_parameters_ = true;
+
+    // Allreduce locally used param maps to get the global consensus
+    local_used_work_ = process_group_->allreduce(local_used_maps_);
```

Comment:
IIUC, @pietern has a concern that this might not work for `no_sync` mode, because `no_sync` mode does not call `prepare_for_backward` for intermediate outputs. When finally calling `prepare_for_backward` on the last output, it might have a different set of unused parameters compared to previous intermediate outputs. So the allreduce here might not compute the correct set of globally unused params.

However, it looks like the same problem already exist in today's DDP, where `prepare_for_backward` is only called on the last output. If `find_unused_parameters` is true, it won't correctly detect unused params either. It seems to me adding global unused support here won't break existing use cases? @pietern